jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannelState.java

```diff
@@ -1154,18 +1150,14 @@ public void upgrade()
             if (LOG.isDebugEnabled())
                 LOG.debug("upgrade {}", toStringLocked());
 
-            switch (_state)
-            {
-                case IDLE:
-                    break;
-                default:
-                    throw new IllegalStateException(getStatusStringLocked());
-            }
+            if (_state != State.IDLE)
+                throw new IllegalStateException(getStatusStringLocked());
+            if (_inputState != InputState.IDLE)
+                throw new IllegalStateException(getStatusStringLocked());
```

Comment:
I think the problem is that as soon as we are unready, there is demand registered that will callback the `AsyncContentProvider` that might do a read.  So if we attempt to do an upgrade, then the new connection will not be able to read (demand pending) and that demand will go to the old `HttpInput`

Reply:
After discussion, I think we are fine.

We can only go to non-idle if we set a `ReadListener` and then `upgrade()` within the `service()` method, which is bad API usage, so it will rightfully throw ISE.

If `upgrade()` is called from the `ReadListener`, then it should be fine: for a WebSocket GET upgrade we would call `onAllDataRead()` (there is no content), and be in IDLE state because we are inside a `ReadListener` method.